File: domain/entities.py
"""
Domain Entities — Core business objects of the simulation.

Entities encapsulate validation and core domain logic.
They depend ONLY on value_objects and exceptions from this layer.

Design Patterns:
- SRP: Each entity validates and manages only its own state.
- OCP: Properties are computed, new derived properties can be added
       via subclassing without modifying existing code.
"""

import math
from dataclasses import dataclass, field
from typing import Optional, List

from domain.exceptions import (
    InvalidMaterialError,
    InvalidLaserConfigError,
    InvalidTrajectoryError,
)


# ============================================================================
# Material Entity
# ============================================================================

@dataclass
class Material:
    """
    Thermal and optical properties of a target material.

    Validates all physical constraints on construction.
    """

    name: str

    # Optical
    absorption: float  # Fraction absorbed (0.0–1.0)

    # Thermal
    thermal_conductivity: float   # k [W/(m·K)]
    density: float                # ρ [kg/m³]
    specific_heat: float          # c_p [J/(kg·K)]

    # Phase change temperatures [°C]
    t_ambient: float = 20.0
    t_melt: float = 1400.0
    t_vaporization: float = 2800.0

    def __post_init__(self):
        self._validate()

    def _validate(self):
        """Enforce physical constraints (Integrity — CIA Triad)."""
        if not (0.0 < self.absorption <= 1.0):
            raise InvalidMaterialError(
                f"Absorption must be in (0, 1], got {self.absorption}"
            )
        if self.thermal_conductivity <= 0:
            raise InvalidMaterialError("Thermal conductivity must be positive.")
        if self.density <= 0:
            raise InvalidMaterialError("Density must be positive.")
        if self.specific_heat <= 0:
            raise InvalidMaterialError("Specific heat must be positive.")
        if self.t_melt >= self.t_vaporization:
            raise InvalidMaterialError(
                f"Melt temp ({self.t_melt}°C) must be below "
                f"vaporization temp ({self.t_vaporization}°C)."
            )

    @property
    def thermal_diffusivity(self) -> float:
        """κ = k / (ρ · c_p) [m²/s]."""
        return self.thermal_conductivity / (self.density * self.specific_heat)


# ============================================================================
# Laser Beam Entity
# ============================================================================

@dataclass
class LaserBeam:
    """
    Laser beam configuration with physics-derived properties.

    Validates parameters on construction.
    """

    power: float            # [W]
    wavelength: float       # [m]
    spot_radius: float      # w0 at focus [m]
    focal_length: float     # [m]

    def __post_init__(self):
        self._validate()

    def _validate(self):
        """Enforce physical constraints."""
        if self.power <= 0:
            raise InvalidLaserConfigError("Power must be positive.")
        if self.wavelength <= 0:
            raise InvalidLaserConfigError("Wavelength must be positive.")
        if self.spot_radius <= 0:
            raise InvalidLaserConfigError("Spot radius must be positive.")
        if self.focal_length <= 0:
            raise InvalidLaserConfigError("Focal length must be positive.")

    @property
    def rayleigh_range(self) -> float:
        """z_R = π · w0² / λ [m]."""
        return math.pi * self.spot_radius ** 2 / self.wavelength

    @property
    def peak_intensity(self) -> float:
        """I_0 = 2P / (π · w0²) [W/m²]."""
        return 2 * self.power / (math.pi * self.spot_radius ** 2)

    @property
    def divergence_angle(self) -> float:
        """θ = λ / (π · w0) [rad]."""
        return self.wavelength / (math.pi * self.spot_radius)


# ============================================================================
# Trajectory Entity
# ============================================================================

@dataclass
class Trajectory:
    """
    Motion path data for laser scanning.

    Stores parallel arrays of time, x, y, z coordinates.
    Validates consistency on construction.
    """

    time: list = field(default_factory=list)
    x: list = field(default_factory=list)
    y: list = field(default_factory=list)
    z: list = field(default_factory=list)

    def __post_init__(self):
        self._validate()

    def _validate(self):
        lengths = {len(self.time), len(self.x), len(self.y), len(self.z)}
        if len(lengths) != 1:
            raise InvalidTrajectoryError(
                "All trajectory arrays must have the same length."
            )
        if len(self.time) < 2:
            raise InvalidTrajectoryError(
                "Trajectory must have at least 2 points."
            )

    @property
    def n_points(self) -> int:
        return len(self.time)

    @property
    def duration(self) -> float:
        return self.time[-1] - self.time[0]


# ============================================================================
# Simulation Result Entity
# ============================================================================

@dataclass
class SimulationResult:
    """
    Immutable container for simulation output.

    Stores the 3D temperature field, fluence map, and derived metrics.
    """

    # Raw data (stored as nested lists for domain purity — no numpy dependency)
    temperature_field: list = field(default_factory=list)
    fluence_map: list = field(default_factory=list)
    x_coords: list = field(default_factory=list)
    y_coords: list = field(default_factory=list)
    z_coords: list = field(default_factory=list)
    time_coords: list = field(default_factory=list)

    # Derived metrics
    peak_temperature_celsius: float = 0.0
    peak_fluence_j_per_m2: float = 0.0
    melt_depth_m: Optional[float] = None
    vaporization_depth_m: Optional[float] = None
    total_energy_j: float = 0.0

    # Metadata
    material_name: str = ""
    solver_name: str = ""
    duration_seconds: float = 0.0

File: UnityHVOF/Assets/Scripts/Domain/Entities/Entities.cs
// ============================================================================
// Domain Entities — Core business objects of the simulation.
//
// Entities encapsulate validation and core domain logic.
// They depend ONLY on ValueObjects and Exceptions from this layer.
//
// Design Patterns:
//   - SRP: Each entity validates and manages only its own state.
//   - OCP: Computed properties can be extended via inheritance.
//
// Port of: domain/entities.py
// ============================================================================

using System;
using System.Collections.Generic;
using HVOFSim.Domain.Exceptions;

namespace HVOFSim.Domain.Entities
{
    // ── Material Entity ──────────────────────────────────────────────

    /// <summary>
    /// Thermal and optical properties of a target material.
    /// Validates all physical constraints on construction (Integrity — CIA Triad).
    /// </summary>
    public sealed class Material
    {
        public string Name { get; }

        // Optical
        public double Absorption { get; }

        // Thermal
        public double ThermalConductivity { get; }
        public double Density { get; }
        public double SpecificHeat { get; }

        // Phase change temperatures [°C]
        public double TAmbient { get; }
        public double TMelt { get; }
        public double TVaporization { get; }

        public Material(
            string name,
            double absorption,
            double thermalConductivity,
            double density,
            double specificHeat,
            double tAmbient = 20.0,
            double tMelt = 1400.0,
            double tVaporization = 2800.0)
        {
            Name = name;
            Absorption = absorption;
            ThermalConductivity = thermalConductivity;
            Density = density;
            SpecificHeat = specificHeat;
            TAmbient = tAmbient;
            TMelt = tMelt;
            TVaporization = tVaporization;

            Validate();
        }

        /// <summary>κ = k / (ρ · c_p) [m²/s].</summary>
        public double ThermalDiffusivity =>
            ThermalConductivity / (Density * SpecificHeat);

        private void Validate()
        {
            if (Absorption <= 0.0 || Absorption > 1.0)
                throw new InvalidMaterialException(
                    $"Absorption must be in (0, 1], got {Absorption}");
            if (ThermalConductivity <= 0)
                throw new InvalidMaterialException("Thermal conductivity must be positive.");
            if (Density <= 0)
                throw new InvalidMaterialException("Density must be positive.");
            if (SpecificHeat <= 0)
                throw new InvalidMaterialException("Specific heat must be positive.");
            if (TMelt >= TVaporization)
                throw new InvalidMaterialException(
                    $"Melt temp ({TMelt}°C) must be below vaporization temp ({TVaporization}°C).");
        }
    }

    // ── Laser Beam Entity ────────────────────────────────────────────

    /// <summary>
    /// Laser beam configuration with physics-derived properties.
    /// Validates parameters on construction.
    /// </summary>
    public sealed class LaserBeam
    {
        public double Power { get; }
        public double Wavelength { get; }
        public double SpotRadius { get; }
        public double FocalLength { get; }

        public LaserBeam(double power, double wavelength, double spotRadius, double focalLength)
        {
            Power = power;
            Wavelength = wavelength;
            SpotRadius = spotRadius;
            FocalLength = focalLength;

            Validate();
        }

        /// <summary>z_R = π · w0² / λ [m].</summary>
        public double RayleighRange =>
            Math.PI * SpotRadius * SpotRadius / Wavelength;

        /// <summary>I_0 = 2P / (π · w0²) [W/m²].</summary>
        public double PeakIntensity =>
            2.0 * Power / (Math.PI * SpotRadius * SpotRadius);

        /// <summary>θ = λ / (π · w0) [rad].</summary>
        public double DivergenceAngle =>
            Wavelength / (Math.PI * SpotRadius);

        private void Validate()
        {
            if (Power <= 0)
                throw new InvalidLaserConfigException("Power must be positive.");
            if (Wavelength <= 0)
                throw new InvalidLaserConfigException("Wavelength must be positive.");
            if (SpotRadius <= 0)
                throw new InvalidLaserConfigException("Spot radius must be positive.");
            if (FocalLength <= 0)
                throw new InvalidLaserConfigException("Focal length must be positive.");
        }
    }

    // ── Trajectory Entity ────────────────────────────────────────────

    /// <summary>
    /// Motion path data for laser scanning.
    /// Stores arrays of time, x, y, z coordinates.
    /// Validates consistency on construction.
    /// </summary>
    public sealed class Trajectory
    {
        public double[] Time { get; }
        public double[] X { get; }
        public double[] Y { get; }
        public double[] Z { get; }

        public Trajectory(double[] time, double[] x, double[] y, double[] z)
        {
            Time = time ?? throw new ArgumentNullException(nameof(time));
            X = x ?? throw new ArgumentNullException(nameof(x));
            Y = y ?? throw new ArgumentNullException(nameof(y));
            Z = z ?? throw new ArgumentNullException(nameof(z));

            Validate();
        }

        public int NPoints => Time.Length;

        public double Duration => Time[Time.Length - 1] - Time[0];

        private void Validate()
        {
            if (Time.Length != X.Length || Time.Length != Y.Length || Time.Length != Z.Length)
                throw new InvalidTrajectoryException(
                    "All trajectory arrays must have the same length.");
            if (Time.Length < 2)
                throw new InvalidTrajectoryException(
                    "Trajectory must have at least 2 points.");
        }
    }

    // ── Simulation Result Entity ─────────────────────────────────────

    /// <summary>
    /// Immutable container for simulation output.
    /// Stores the 3D temperature field, fluence map, and derived metrics.
    /// </summary>
    public sealed class SimulationResult
    {
        // Raw data
        public double[,,] TemperatureField { get; }
        public double[,] FluenceMap { get; }
        public double[] XCoords { get; }
        public double[] YCoords { get; }
        public double[] ZCoords { get; }
        public double[] TimeCoords { get; }

        // Derived metrics
        public double PeakTemperatureCelsius { get; }
        public double PeakFluenceJPerM2 { get; }
        public double? MeltDepthM { get; }
        public double? VaporizationDepthM { get; }
        public double TotalEnergyJ { get; }

        // Metadata
        public string MaterialName { get; }
        public string SolverName { get; }
        public double DurationSeconds { get; }

        public SimulationResult(
            double[,,] temperatureField,
            double[,] fluenceMap,
            double[] xCoords,
            double[] yCoords,
            double[] zCoords,
            double[] timeCoords,
            double peakTemperatureCelsius,
            double peakFluenceJPerM2,
            double? meltDepthM,
            double? vaporizationDepthM,
            double totalEnergyJ,
            string materialName,
            string solverName,
            double durationSeconds)
        {
            TemperatureField = temperatureField;
            FluenceMap = fluenceMap;
            XCoords = xCoords;
            YCoords = yCoords;
            ZCoords = zCoords;
            TimeCoords = timeCoords;
            PeakTemperatureCelsius = peakTemperatureCelsius;
            PeakFluenceJPerM2 = peakFluenceJPerM2;
            MeltDepthM = meltDepthM;
            VaporizationDepthM = vaporizationDepthM;
            TotalEnergyJ = totalEnergyJ;
            MaterialName = materialName;
            SolverName = solverName;
            DurationSeconds = durationSeconds;
        }
    }
}
